state-and-event-handling (#38)
Ada-Developers-Academy/core-unit-3: state-and-event-handling/event-handling.md

Consider this code, where `printMessage` accepts an event object, and then prints it.

<!-- prettier-ignore-start -->
```js
const Post = () => {
    const printMessage = (event) => {
        console.log('Hello! We\'re in printMessage!');
        console.log('This event object contains details about the event:', event);
    }

    return (<section>
        <button onClick={printMessage}>Like</button>
    </section>)
}
```
<!-- prettier-ignore-end -->

This is what prints to the browser console. (Your results may vary depending on your browser.)

```
Hello! We're in printMessage!
This event object contains details about the event: 
Object { _reactName: "onClick", ...
```

### Event Handling Functions Are Functions

It's helpful to know that our event handling functions are, otherwise, normal functions.

They can call other functions inside them, and pass arguments to them, too.

Consider this `Post` component. It:

- Defines a function, `printName`, which takes in a `name` argument
- Defines another function, `printMessage`. It calls `printName` with the argument `"Ada Lovelace"`.
- Sets `printMessage` as the click event handler for a "Like" button

<!-- prettier-ignore-start -->

> ```js

Reviewer: I really want to bring up closure binding again. I think it's ok to skip it here, since if we squint at this, we can convince ourselves that we're just looking at a funny looking class. But eventually closure semantics wind up rearing their head and ordering can become an issue (especially with useCallback dependencies).

Author: YEAHHHHHH, I just finished writing the `lifting-state-up` PRs and I really craved this.

I'll copy/paste this into Slack for a fun convo, buuUUUT I had made the call that:

- Closures are a lightning talk topic
- Specifically for C15, I know that one of the Closure lightning talk talks with be late 🙈 Not ideal...

Author: I really don't want to forget this, so I made an issue for it to help me remember: https://github.com/Ada-Developers-Academy/core-unit-3/issues/44

const Post = () => {
    const printName = (name) => {
        console.log('We\'re in printName!');
        console.log(`Hello, ${name}!`);
    }

    const printMessage = () => {
        printName('Ada Lovelace');
        console.log('Now, we\'re in printMessage!');
    }

    return (<section>
        <button onClick={printMessage}>Like</button>
    </section>)
}
```
<!-- prettier-ignore-end -->

Our output proves that our `printMessage` event handler can invoke other functions.

```
We're in printName!
Hello, Ada Lovelace!
Now, we're in printMessage!
```

## Pattern: Anonymous Functions

Our event handlers can be _anonymous functions_.

Consider our `printMessage` and `printName` example. Let's refactor our `printMessage` into an inline, anonymous function:

<!-- prettier-ignore-start -->
```js

> const Post = () => {
>     const printName = (name) => {
>         console.log('We\'re in printName!');
>         console.log(`Hello, ${name}!`);
>     }
> 
>     return (<section>
>         <button onClick={() => printName('Ada Lovelace')}>Like</button>
>     </section>)
> }

Reviewer: spacing and ;
```suggestion
const Post = () => {
    const printName = (name) => {
        console.log('We\'re in printName!');
        console.log(`Hello, ${name}!`);
    };

    return (
        <section>
            <button onClick={() => printName('Ada Lovelace')}>Like</button>
        </section>
    );
};
```

```
<!-- prettier-ignore-end -->

This code produces the output:

```
We're in printName!
Hello, Ada Lovelace!
```

Every time we click on our "like" button, our anonymous function will fire. This, in turn, invokes `printName('Ada Lovelace')`!

> This is a common pattern in React code.

Reviewer: I don't think this is clear about the pattern. Let's be a little more explicit and warn about a common error students run into.
```suggestion
Using anonymous functions for event handlers is a common pattern in React, especially when the function we want to call needs us to pass in some parameters.

A common mistake when trying to call a function that takes an argument as an event handler is to try to write the function call, in this case `printName('Ada Lovelace')`, directly as the even handler.

But this does not have the desired result! Instead, it:
1. Calls `printName('Ada Lovelace')` as the button is being rendered rather than when it is clicked. This causes the output message to appear while the page is still being rendered. Not what we wanted!
1. Sets the _result_ of calling `printName('Ada Lovelace')` as the click handler. Since `printName` returns nothing, the result is `undefined`. So no event handler is registered for `onClick`, and the button does nothing when it is clicked. Also not what we wanted!
```

> When our event handler has at least one parameter, we can use this technique.

Reviewer: expand this part of the explanation too. recognize this pattern works well in some situations, but named functions still work better in others.
```suggestion
So when our desired behavior takes at least one parameter, we can use this technique of wrapping it in an anonymous function.

An anonymous function lets us call arbitrary code exactly where the event registration occurs, which can be good for understandability. But we may find that, if we need to do more than one or two things in the event handler, then using a named function may still be more understandable.
```
